output/cloud/expv2/collect.go

```diff
@@ -0,0 +1,158 @@
+		}
+		m[s.TimeSeries] = b
+	}
+	b.Sink.Add(s)
+}
+
+type collector struct {
+	buf   *output.SampleBuffer
+	bq    bucketQ
+	nowfn func() time.Time
+
+	aggregationPeriod time.Duration
+	waitPeriod        time.Duration
+
+	// we should no longer have to handle metrics that have times long in the past. So instead of a
+	// map, we can probably use a simple slice (or even an array!) as a ring buffer to store the
+	// aggregation buckets. This should save us a some time, since it would make the lookups and WaitPeriod
+	// checks basically O(1). And even if for some reason there are occasional metrics with past times that
+	// don't fit in the chosen ring buffer size, we could just send them along to the buffer unaggregated
+	aggrBuckets map[int64]perSeriesBuckets
+}
+
+func newCollector(buf *output.SampleBuffer, aggrPeriod, waitPeriod time.Duration) (*collector, error) {
+	if aggrPeriod == 0 {
+		return nil, errors.New("aggregation period is not allowed to be zero")
+	}
+	return &collector{
+		buf:               buf,
+		bq:                bucketQ{},
+		nowfn:             time.Now,
+		aggrBuckets:       make(map[int64]perSeriesBuckets),
+		aggregationPeriod: aggrPeriod,
+		waitPeriod:        waitPeriod,
+	}, nil
+}
+
+// CollectSamples drain the buffer and collect all the samples.
+func (c *collector) CollectSamples() {
+	c.bucketing(c.buf.GetBufferedSamples())
+
+	expired := c.expiredBuckets()
+	if len(expired) > 0 {
+		c.bq.Push(expired)
+	}
+}
+
+func (c *collector) bucketID(t time.Time) int64 {
+	return t.UnixNano() / int64(c.aggregationPeriod)
+}
+
+func (c *collector) bucketing(containers []metrics.SampleContainer) {
+	// Distribute all newly buffered samples into related buckets
+	for _, sampleContainer := range containers {
+		samples := sampleContainer.GetSamples()
+
+		for i := 0; i < len(samples); i++ {
+			bucketID := c.bucketID(samples[i].Time)
+
+			// Get or create a time bucket
+			buckets, ok := c.aggrBuckets[bucketID]
+			if !ok {
+				buckets = newPerSeriesBuckets()
+				c.aggrBuckets[bucketID] = buckets
+			}
+
+			buckets.AddSample(bucketID, samples[i], int64(c.aggregationPeriod))
+		}
+	}
+}
+
+func (c *collector) expiredBuckets() []timeBucket {
+	// Still new buckets where we have to wait to accumulate
+	// more samples before flushing
+	//
+	// TODO: Do we need it?
+	// if the Cloud handles the merge is not really required
+	bucketCutoffID := c.nowfn().Add(-c.waitPeriod).UnixNano() / int64(c.aggregationPeriod)
+
+	// TODO: multiply it at least for the number of metrics?
+	expired := make([]timeBucket, 0, len(c.aggrBuckets))
+
+	// Handle all aggregation buckets older than bucketCutoffID
+	for bucketID, seriesBuckets := range c.aggrBuckets {
+		if bucketID > bucketCutoffID {
+			continue
+		}
+
+		for _, bucket := range seriesBuckets {
+			// TODO: is there a valid math reason to have a minimum
+			// number of samples aggregated as we do in v1?
+			// I mean o.config.AggregationMinSamples.Int64
+
```

Comment:
Given that the aggregation happens as you add them - I don't think it is even possible :).

Reply:
I think we could count the number of samples per bucket. In any case, I don't think we have a valid use case to justify this overhead.